File: framework/backend/go/k1s0-domain-event/outbox/entry.go
// Package outbox provides the transactional outbox pattern for domain events.
package outbox

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

// Status represents the state of an outbox entry.
type Status string

const (
	StatusPending   Status = "pending"
	StatusPublished Status = "published"
	StatusFailed    Status = "failed"
)

// Entry represents a single row in the outbox table.
type Entry struct {
	ID          string          `json:"id"`
	EventType   string          `json:"event_type"`
	Payload     json.RawMessage `json:"payload"`
	Status      Status          `json:"status"`
	CreatedAt   time.Time       `json:"created_at"`
	PublishedAt *time.Time      `json:"published_at,omitempty"`
	RetryCount  int             `json:"retry_count"`
	LastError   string          `json:"last_error,omitempty"`
}

// NewEntry creates a new pending outbox entry.
func NewEntry(eventType string, payload json.RawMessage) *Entry {
	return &Entry{
		ID:         uuid.New().String(),
		EventType:  eventType,
		Payload:    payload,
		Status:     StatusPending,
		CreatedAt:  time.Now().UTC(),
		RetryCount: 0,
	}
}

File: framework/backend/python/packages/k1s0-domain-event/src/k1s0_domain_event/outbox/entry.py
"""Outbox entry model for the transactional outbox pattern."""

from __future__ import annotations

from datetime import UTC, datetime
from enum import Enum
from uuid import UUID, uuid4

from pydantic import BaseModel, Field


class OutboxStatus(str, Enum):
    """Status of an outbox entry."""

    PENDING = "pending"
    PUBLISHED = "published"
    FAILED = "failed"


class OutboxEntry(BaseModel):
    """Represents a single entry in the outbox table."""

    id: UUID = Field(default_factory=uuid4)
    event_type: str
    payload: str
    status: OutboxStatus = OutboxStatus.PENDING
    retry_count: int = 0
    created_at: datetime = Field(default_factory=lambda: datetime.now(UTC))
    updated_at: datetime = Field(default_factory=lambda: datetime.now(UTC))
